README.md
# Group Grid View
A GridView that allows you to group list items and support headers like iOS UICollectionView section.

## Features

* Grid Items can be grouped.
* Support headers and footers for each group.
* All fields from GridView.builder constructor available.

  ![Alt Text](https://images.viblo.asia/5233b34d-8fc3-47f1-91ff-e31929fbc2a9.gif)
## Installing

Add this to your package's `pubspec.yaml` file:

```dart
dependencies:
  group_grid_view: ^latest
```

## Import

```dart
import 'package:group_grid_view/group_grid_view.dart';
```

## Usage

Use like a widget everywhere you want, you can see example in `/example` folder.

```dart
GroupGridView(
  padding: const EdgeInsets.symmetric(horizontal: 16),
  gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
    crossAxisCount: 3, mainAxisSpacing: 16, crossAxisSpacing: 16),
  sectionCount: dataSource.length,
  headerForSection: (section) => Container(
    padding: const EdgeInsets.symmetric(vertical: 12),
    child: Text(dataSource[section].sectionName,
      style: const TextStyle(fontSize: 22, fontWeight: FontWeight.bold))),
  footerForSection: (section) {
    final footer = dataSource[section].footerName;
    return footer != null
          ? Container(
              padding: const EdgeInsets.only(top: 12, bottom: 32),
              child: Text(footer, style: const TextStyle(fontSize: 16)))
          : const SizedBox(height: 32);
  },
  itemInSectionBuilder: (_, indexPath) {
    final data = dataSource[indexPath.section].listItems[indexPath.index];
    return Container(
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(8),
        border: Border.all(color: Colors.blue)),
      alignment: Alignment.center,
      padding: const EdgeInsets.all(8),
      child: Text(data, style: const TextStyle(fontWeight: FontWeight.w500)));
  },
  itemInSectionCount: (section) => dataSource[section].listItems.length)
```


**Trần Đạt, Thank you!**

CHANGELOG.md
## 1.0.0

* Publish first version of package.


lib/src/group_grid_view.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

class GroupGridView extends StatefulWidget {
  const GroupGridView(
      {required this.gridDelegate,
      required this.sectionCount,
      required this.itemInSectionBuilder,
      required this.itemInSectionCount,
      this.headerForSection,
      this.footerForSection,
      this.scrollDirection = Axis.vertical,
      this.reverse = false,
      this.controller,
      this.primary,
      this.physics,
      this.shrinkWrap = false,
      this.padding,
      this.findChildIndexCallback,
      this.addAutomaticKeepAlive = true,
      this.addRepaintBoundaries = true,
      this.addSemanticIndexes = true,
      this.cacheExtent,
      this.dragStartBehavior = DragStartBehavior.start,
      this.keyboardDismissBehavior = ScrollViewKeyboardDismissBehavior.manual,
      this.restorationId,
      this.clipBehavior = Clip.hardEdge,
      super.key});

  final SliverGridDelegate gridDelegate;
  final int sectionCount;
  final Widget Function(BuildContext context, IndexPath indexPath)
      itemInSectionBuilder;
  final int Function(int section) itemInSectionCount;

  final Widget? Function(int section)? headerForSection;
  final Widget? Function(int section)? footerForSection;

  final Axis scrollDirection;
  final bool reverse;
  final ScrollController? controller;
  final bool? primary;
  final ScrollPhysics? physics;
  final bool shrinkWrap;
  final EdgeInsetsGeometry? padding;
  final IndexPath? Function(Key)? findChildIndexCallback;
  final bool addAutomaticKeepAlive;
  final bool addRepaintBoundaries;
  final bool addSemanticIndexes;
  final double? cacheExtent;
  final DragStartBehavior dragStartBehavior;
  final ScrollViewKeyboardDismissBehavior keyboardDismissBehavior;
  final String? restorationId;
  final Clip clipBehavior;

  @override
  State<StatefulWidget> createState() => _GroupGridViewState();
}

class _GroupGridViewState extends State<GroupGridView> {
  @override
  Widget build(BuildContext context) {
    return Padding(
        padding: widget.padding ?? EdgeInsets.zero,
        child: CustomScrollView(
            scrollDirection: widget.scrollDirection,
            reverse: widget.reverse,
            controller: widget.controller,
            primary: widget.primary,
            physics: widget.physics,
            shrinkWrap: widget.shrinkWrap,
            cacheExtent: widget.cacheExtent,
            dragStartBehavior: widget.dragStartBehavior,
            keyboardDismissBehavior: widget.keyboardDismissBehavior,
            restorationId: widget.restorationId,
            clipBehavior: widget.clipBehavior,
            slivers: _listSliver));
  }

  List<Widget> get _listSliver {
    List<Widget> list = [];
    for (var section = 0; section < widget.sectionCount; section++) {
      final header = widget.headerForSection?.call(section);
      if (header != null) {
        list.add(SliverToBoxAdapter(child: header));
      }

      list.add(SliverGrid.builder(
          addAutomaticKeepAlives: widget.addAutomaticKeepAlive,
          addRepaintBoundaries: widget.addRepaintBoundaries,
          addSemanticIndexes: widget.addSemanticIndexes,
          gridDelegate: widget.gridDelegate,
          itemBuilder: (context, index) => widget.itemInSectionBuilder
              .call(context, IndexPath(section: section, index: index)),
          itemCount: widget.itemInSectionCount.call(section)));

      final footer = widget.footerForSection?.call(section);
      if (footer != null) {
        list.add(SliverToBoxAdapter(child: footer));
      }
    }
    return list;
  }
}

class IndexPath {
  const IndexPath({required this.section, required this.index});

  final int section;
  final int index;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:group_grid_view/group_grid_view.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue
      ),
      home: const MyHomePage(title: 'Group Grid View'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final List<TestModel> dataSource = [
    TestModel(
        sectionName: "Animals",
        listItems: ["Dog", "Cat", "Bear", "Tiger", "Lion"]),
    TestModel(
        sectionName: "Vehicle",
        footerName: "End of vehicle section",
        listItems: ["Car", "Motor", "Air plane", "Boat"]),
    TestModel(sectionName: "Countries", listItems: [
      "Vietnam",
      "China",
      "Laos",
      "Thai",
      "Japan",
      "Singapore",
      "England",
      "USA",
      "Russian"
    ])
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: GroupGridView(
          padding: const EdgeInsets.symmetric(horizontal: 16),
          gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
              crossAxisCount: 3, mainAxisSpacing: 16, crossAxisSpacing: 16),
          sectionCount: dataSource.length,
          headerForSection: (section) => Container(
              padding: const EdgeInsets.symmetric(vertical: 12),
              child: Text(dataSource[section].sectionName,
                  style: const TextStyle(
                      fontSize: 22, fontWeight: FontWeight.bold))),
          footerForSection: (section) {
            final footer = dataSource[section].footerName;
            return footer != null
                ? Container(
                    padding: const EdgeInsets.only(top: 12, bottom: 32),
                    child: Text(footer, style: const TextStyle(fontSize: 16)))
                : const SizedBox(height: 32);
          },
          itemInSectionBuilder: (_, indexPath) {
            final data =
                dataSource[indexPath.section].listItems[indexPath.index];
            return Container(
                decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(8),
                    border: Border.all(color: Colors.blue)),
                alignment: Alignment.center,
                padding: const EdgeInsets.all(8),
                child: Text(data, style: const TextStyle(fontWeight: FontWeight.w500)));
          },
          itemInSectionCount: (section) => dataSource[section]
              .listItems
              .length)
    );
  }
}

class TestModel {
  TestModel(
      {required this.sectionName, required this.listItems, this.footerName});

  final String sectionName;
  final String? footerName;
  final List<String> listItems;
}
